File: day-04/part2.go
package main

import (
	"github.com/dawidl022/advent-of-code-2021/day-04/logic"
	"github.com/dawidl022/advent-of-code-2021/day-04/util"
)

func solution2(numbers []int, boards []*logic.Board) int {
	countOfWonBoards := 0

	for _, number := range numbers {
		for _, board := range boards {
			if board.Won {
				continue
			}

			board.MarkNumberIfPresent(number)

			if board.HasWinningLine() {
				board.Won = true
				countOfWonBoards++
			}

			if countOfWonBoards == len(boards) {
				return number * board.SumOfUnmarkedNumbers()
			}
		}
	}
	util.Terminate(notAllBoardsWonError{})
	return -1
}

type notAllBoardsWonError struct{}

func (e notAllBoardsWonError) Error() string {
	return "not all boards won from given input"
}

func main() {
	util.Run(solution2)
}

File: day-04/part2.py
from util.run import run
from util.types import Board
from logic.logic import mark_number_if_present, has_winning_row, sum_of_unmarked_numbers


def solution(numbers: list[int], boards: list[Board]) -> int:
    won_boards_indices = set()

    for number in numbers:
        for i in range(len(boards)):
            if i in won_boards_indices:
                continue

            board = boards[i]
            mark_number_if_present(number, board)

            if has_winning_row(board):
                won_boards_indices.add(i)

            if len(won_boards_indices) == len(boards):
                return number * sum_of_unmarked_numbers(board)

    raise NotAllBoardsWonException()


class NotAllBoardsWonException(Exception):
    pass


if __name__ == "__main__":
    run(solution)
